mypy/checkexpr.py

```diff
@@ -2293,12 +2293,15 @@ def visit_comparison_expr(self, e: ComparisonExpr) -> Type:
                     self.msg.add_errors(local_errors)
             elif operator in operators.op_methods:
                 method = self.get_operator_method(operator)
-                err_count = self.msg.errors.total_errors()
-                sub_result, method_type = self.check_op(method, left_type, right, e,
-                                                        allow_reverse=True)
+
+                with ErrorWatcher(self.msg.errors) as w:
```

Comment:
There is a similar mechanism being used in the `infer_overload_return_type` method in this file (grep for `.clean_copy`). Can we use that here too?

Reply:
Hmm. It seems like `infer_overload_return_type` creates a new *empty* `MessageBuilder` / `Errors` and uses that to check if any new errors where triggered, which could be achieved by using the new `ErrorWatcher`. However it also looks like any such errors would then be ignored (not actually make it into the original `Errors`), which the `ErrorWatcher` approach alone cannot do, and if we're creating a new empty `Errors` then there's not much benefit to the `ErrorWatcher` approach. Am I missing something?

Reply:
Could we use the `infer_overload_return_type` approach here, so we don't have to introduce the new `ErrorWatcher` concept? Seems like we could do that here by getting a clean copy, then unconditionally doing `self.msg.add_errors` like on line 2293 in visit_comparison_expr.

Reply:
hmm, I think the current approach there needs to be rethought.

Consider on line 1757:
``` 
            self.msg = overload_messages
            self.chk.msg = overload_messages
            try:
                # Passing `overload_messages` as the `arg_messages` parameter doesn't
                # seem to reliably catch all possible errors.
                # TODO: Figure out why
```

which seems obviously pretty related to the way  the `MessageBuilder` in `TypeChecker` is created:

```
        self.msg = MessageBuilder(errors, modules)
        self.plugin = plugin
        self.expr_checker = mypy.checkexpr.ExpressionChecker(self, self.msg, self.plugin)
        self.pattern_checker = PatternChecker(self, self.msg, self.plugin)
```

The explicit patching of the `MessageBuilder` instance is incomplete (expr_checker and pattern_checker are not patched), but more importantly it is very fragile (there might be other places that need patching, the places that need patching are likely to change over time in ways that hard hard to keep track off), and rather ugly to boot. I think instead of using this pattern in more places we should replace it with some variant of `ErrorWatcher` that can optionally intercept/discard errors in the original `MessageBuilder`

Reply:
That's a good point. I suspect bug #12665, which we found yesterday, is related to the problem you identify.